### PR TITLE
[Doc] GPU custom ops tutorial updates.

### PR DESCRIPTION
**Pull request by nouiz**
Add reference to the C code where I was looking for them.
Also add some high-level description of what is needed.

@seanprime7 @sharadmv

### CODE REVIEW

File: docs/Custom_Operation_for_GPUs.md [hunk 1]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
 
 To accommodate such scenarios, JAX allows users to define custom operations and this tutorial is to explain how we can define one for GPUs and use it in single-GPU and multi-GPU environments.
 
-This tutorial contains information from [Extending JAX with custom C++ and CUDA code](https://github.com/dfm/extending-jax).
+This tutorial contains information from [Extending JAX with custom C++ and CUDA code](https://github.com/dfm/extending-jax) and
+suppose that you are familiar with [JAX primitive](https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html).
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:8]: supposes

File: docs/Custom_Operation_for_GPUs.md [hunk 2]
```diff
@@ -85,6 +112,7 @@ PYBIND11_MODULE(gpu_ops, m) {
 ## Build `gpu_ops` extension module
 
 We build the `gpu_ops` Python extension module with the aforementioned code.
+(See gpu_ops code listing for complete code listing of C++ and CUDA files.)
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:115]: gpu_ops code listing needs to be formatted

File: docs/Custom_Operation_for_GPUs.md [hunk 3]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
 
 For this tutorial, we are going to add the RMS normalization as a custom operation in JAX.
 Note that the RMS normalization can be expressed with [`jax.numpy`](https://jax.readthedocs.io/en/latest/jax.numpy.html) directly. However, we are using it as an example to show the process of creating a custom operation for GPUs.
 The CUDA code in `gpu_ops/rms_norm_kernels.cu` for this operation has been borrowed from [Apex](https://github.com/NVIDIA/apex/blob/master/csrc/layer_norm_cuda_kernel.cu) and adapted to eliminate any dependency on PyTorch.
-See [`gpu_ops` code listing](#gpu_ops-code-listing) for complete code listing of C++ and CUDA files.
 
+
+## High-level:
+
+This tutorial show how to have custom operation for the forward and gradient of an operation.
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:19]: shows

Comment (seanprime7) [reply docs/Custom_Operation_for_GPUs.md:19]: custom operation for … of an operation sounds a little unnatural.  Could we re-write it?

File: docs/Custom_Operation_for_GPUs.md [hunk 4]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+
+In C:
+You need to do those steps in C for each new JAX primitive:
+* CUDA kernel(s).
+* A C function that dispatch the CUDA kernel, it will be called by XLA.
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:24]: dispatches

File: docs/Custom_Operation_for_GPUs.md [hunk 5]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+* A descriptor to convey information needed for the computation.
+  * The type(s), the shapes and other attributes.
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:26]: type(s) but not shape(s)?  I think we can skip parentheses for both

File: docs/Custom_Operation_for_GPUs.md [hunk 6]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
 
 ## RMS normalization
 
 For this tutorial, we are going to add the RMS normalization as a custom operation in JAX.
 Note that the RMS normalization can be expressed with [`jax.numpy`](https://jax.readthedocs.io/en/latest/jax.numpy.html) directly. However, we are using it as an example to show the process of creating a custom operation for GPUs.
 The CUDA code in `gpu_ops/rms_norm_kernels.cu` for this operation has been borrowed from [Apex](https://github.com/NVIDIA/apex/blob/master/csrc/layer_norm_cuda_kernel.cu) and adapted to eliminate any dependency on PyTorch.
-See [`gpu_ops` code listing](#gpu_ops-code-listing) for complete code listing of C++ and CUDA files.
 
+
+## High-level:
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:17]: Can we remove the colon at the end as it is a heading?

Can we change it to something like High-level steps?

File: docs/Custom_Operation_for_GPUs.md [hunk 7]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+
+This tutorial show how to have custom operation for the forward and gradient of an operation.
+
+In C:
+You need to do those steps in C for each new JAX primitive:
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:22]: “follow these steps” or “follow the steps below” instead of “do those steps” would make it clearer

File: docs/Custom_Operation_for_GPUs.md [hunk 8]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+* CUDA kernel(s).
+* A C function that dispatch the CUDA kernel, it will be called by XLA.
+* A descriptor to convey information needed for the computation.
+  * The type(s), the shapes and other attributes.
+* Bind C functions to Python
+  * To create the descriptor and to call the primitive during execution.
+
+In Python:
+* You need to do those steps in Python:
```
Comment (seanprime7) [inline docs/Custom_Operation_for_GPUs.md:31]: “follow these steps” or “follow the steps below” instead of “do those steps” would make it clearer

File: docs/Custom_Operation_for_GPUs.md (hunk 3 above)
Comment (sharadmv) [reply docs/Custom_Operation_for_GPUs.md:19]: Perhaps: "This tutorial shows how to write both a custom operation and its gradient."

File: docs/Custom_Operation_for_GPUs.md (hunk 4 above)
Comment (sharadmv) [reply docs/Custom_Operation_for_GPUs.md:24]: Suggestion: "A C function that dispatches the CUDA kernel that will be called by XLA"

File: docs/Custom_Operation_for_GPUs.md (hunk 7 above)
Comment (sharadmv) [reply docs/Custom_Operation_for_GPUs.md:22]: For each step, could you add a verb (i.e. "Write a CUDA kernel" and "Write a C function") to make it more consistent with "follow these steps"?

File: docs/Custom_Operation_for_GPUs.md [hunk 9]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+* Define a new JAX primitive (instruction/operation)
+* Python functions to build the graph nodes with the primitive.
```
Comment (sharadmv) [inline docs/Custom_Operation_for_GPUs.md:33]: "Write Python functions"

File: docs/Custom_Operation_for_GPUs.md [hunk 10]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+* Define its abstract evaluation.
+* Define its lowering to MLIR.
+* [Optional] Define the gradient.
+* [Optional] Use xmap (or the experimental custom_partitioning) for fast multi-GPU.
```
Comment (sharadmv) [inline docs/Custom_Operation_for_GPUs.md:37]: Perhaps mention `shard_map` here as well?

File: docs/Custom_Operation_for_GPUs.md [hunk 11]
```diff
@@ -4,15 +4,42 @@ JAX ships with a large number of built-in operations, but users occasionally run
...
+
+
+## C code
+
+See [`gpu_ops` code listing](#gpu_ops-code-listing) for complete code listing of C++ and CUDA files.
```
Comment (sharadmv) [inline docs/Custom_Operation_for_GPUs.md:42]: Nit: "for *a* complete code listing"

**Review by sharadmv — COMMENTED** [review]
Mostly grammatical nits

File: docs/Custom_Operation_for_GPUs.md (hunk 10 above)
Comment (nouiz) [reply docs/Custom_Operation_for_GPUs.md]: Added, and I added links.

File: docs/Custom_Operation_for_GPUs.md (hunk 7 above)
Comment (nouiz) [reply docs/Custom_Operation_for_GPUs.md]: Good idea. I did it for the Python listing too.

**Comment by nouiz** [comment]
I fixed all the comments.

**Comment by jakevdp** [comment]
Thanks! Can you squash the changes into a single commit? (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests) Thanks!

**Comment by nouiz** [comment]
done

**Review by jakevdp — APPROVED** [review]
Thanks!
